hw/xfree86/drivers/video/modesetting/drmmode_display.c

```diff
@@ -1140,24 +1152,31 @@ drmmode_create_bo(drmmode_ptr drmmode, drmmode_bo *bo,
         }
 
 #ifdef GBM_BO_WITH_MODIFIERS
-        num_modifiers = get_modifiers_set(drmmode->scrn, format, &modifiers,
-                                          FALSE, TRUE);
-        if (num_modifiers > 0 &&
-            !(num_modifiers == 1 && modifiers[0] == DRM_FORMAT_MOD_INVALID)) {
-            bo->gbm = gbm_bo_create_with_modifiers(drmmode->gbm, width, height,
-                                                   format, modifiers,
-                                                   num_modifiers);
-            free(modifiers);
-            if (bo->gbm) {
-                bo->used_modifiers = TRUE;
-                return TRUE;
-            }
-        }
+        uint64_t *modifiers;
+        uint32_t num_modifiers =
+                get_modifiers_set(drmmode->scrn, format, &modifiers, FALSE, TRUE, TRUE);
+
+#ifdef GBM_BO_WITH_MODIFIERS2
+        bo->gbm = gbm_bo_create_with_modifiers2(drmmode->gbm, 
+                                                width, height, format,
+                                                modifiers, num_modifiers,
+                                                GBM_BO_USE_RENDERING | GBM_BO_USE_SCANOUT);
+#else
+        /* gbm_bo_create_with_modifiers assumes GBM_BO_USE_SCANOUT */
+        bo->gbm = gbm_bo_create_with_modifiers(drmmode->gbm, 
+                                               width, height, format,
+                                               modifiers, num_modifiers);
 #endif
-
+        bo->used_modifiers = (bo->gbm && num_modifiers);
+        
+        if (modifiers)
+            free(modifiers);
+#else
```

Comment:
This needs to be:
```suggestion
        if (!bo->gbm) {
```
Otherwise, we're no longer creating a bo in the case with no modifiers.